pilot/pkg/model/jwks_resolver.go

```diff
@@ -42,20 +44,76 @@ const (
 	// JwksURI Cache eviction time duration, cache eviction is done on a periodic basis,
 	// jwksURICacheEviction specifies the frequency at which eviction activities take place.
 	jwksURICacheEviction = time.Minute * 30
+
+	// JwtPubKeyExpireDuration is the expire duration for JWT public key in the cache.
+	// After this duration expire, refresher job will fetch key for the cached item again.
+	JwtPubKeyExpireDuration = time.Hour
+
+	// JwtPubKeyEvictionDuration is the life duration for cached item.
+	// Cached item will be removed from the cache if it hasn't been used longer than JwtPubKeyEvictionDuration.
+	JwtPubKeyEvictionDuration = 24 * 7 * time.Hour
+
+	// JwtPubKeyRefreshInterval is the running interval of JWT pubKey refresh job.
+	JwtPubKeyRefreshInterval = time.Minute * 20
 )
 
-// jwksURIResolver is wrapper of expiringCache for jwksUri.
-type jwksURIResolver struct {
-	cache cache.ExpiringCache
+// jwtPubKeyEntry is a single cached entry for jwt public key.
+type jwtPubKeyEntry struct {
+	pubKey string
+
+	// Cached item will be fetched again by refresher job if (time.now >= expireTime).
+	expireTime time.Time
+
+	// Cached item's last used time, which is set in GetPublicKey.
+	lastUsedTime time.Time
 }
 
-// newJwksURIResolver creates new instance of jwksURIResolver.
-func newJwksURIResolver() jwksURIResolver {
-	return jwksURIResolver{cache.NewTTL(jwksURICacheExpiration, jwksURICacheEviction)}
+// jwksResolver is resolver for jwksURI and jwt public key.
+type jwksResolver struct {
+	// cache for jwksURI.
+	JwksURICache cache.ExpiringCache
+
+	// cache for JWT public key.
+	// map key is jwksURI, map value is jwtPubKeyEntry.
+	keyEntries sync.Map
+
+	client        *http.Client
+	closing       chan bool
+	refreshTicker *time.Ticker
+
+	expireDuration time.Duration
+
+	// Cached key will be removed from cache if (time.now - cachedItem.lastUsedTime >= evictionDuration), this prevents key cache growing indefinitely.
+	evictionDuration time.Duration
+
+	// Refresher job running interval.
+	refreshInterval time.Duration
+
+	// How may times refresh job has detected JWT public key change happened, used in unit test.
+	keyChangedCount uint64
```

Comment:
If you really care about keyChangeCount - please add a prometheus.Counter here. 

You can add prom for other things as well - cache size, number of requests, error codes, etc.